### PR TITLE
[Console] Sync ConsoleLogger::interpolate with the one in HttpKernel

### PR DESCRIPTION
**Pull request by dunglas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no
| Deprecations? | no 
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | n/a

Adapted from: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/Log/Logger.php#L92

Better performance and datetime support (maybe should it be merged in a newer version, but I've targeted 2.7 to prevent future merge conflicts).

### CODE REVIEW

**Review by nicolas-grekas — APPROVED** [review]

**Comment by sroze** [comment]
Wouldn't it be better to share the code instead of copy/pasting it?

**Comment by dunglas** [comment]
Where would be this code? There is no dependency in common between `console` and `http-kernel`. And creating a new composant just for this method looks overkill.

**Comment by nicolas-grekas** [comment]
@sroze these are two independent components, so sharing is not possible without creating news deps, which we don't want to.

**Review by chalasr — APPROVED** [review]

File: src/Symfony/Component/Console/Logger/ConsoleLogger.php [hunk 1]
```diff
@@ -105,15 +105,23 @@ public function log($level, $message, array $context = array())
...
+            return $message;
+        }
+
+        $replacements = array();
         foreach ($context as $key => $val) {
-            if (!is_array($val) && (!is_object($val) || method_exists($val, '__toString'))) {
-                $replace[sprintf('{%s}', $key)] = $val;
+            if (null === $val || is_scalar($val) || (\is_object($val) && method_exists($val, '__toString'))) {
+                $replacements["{{$key}}"] = $val;
+            } elseif ($val instanceof \DateTime || $val instanceof \DateTimeInterface) {
```
Comment (chalasr) [inline src/Symfony/Component/Console/Logger/ConsoleLogger.php]: note to mergers: first condition should be removed starting from 3.3

**Comment by chalasr** [comment]
This seems to break the appveyor build :/

**Comment by fabpot** [comment]
Not a bug fix, should be merged in 3.4 instead.

File: src/Symfony/Component/Console/Logger/ConsoleLogger.php (hunk 1 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Console/Logger/ConsoleLogger.php]: @dunglas can you please rebase+retarget, and take this comment into account meanwhile?

**Comment by dunglas** [comment]
Status: Needs Review

**Comment by chalasr** [comment]
Thank you @dunglas.

**Pull request merged**